utils/writeable/src/fallible.rs

```diff
@@ -0,0 +1,660 @@
+#[derive(Debug)]
+pub struct PanickyWriteable<T>(pub(crate) T);
+
+fn panicky_handler<E>(e: E) -> Result<NeverWriteable, E> {
+    Err(e)
+}
+
+impl<T> Writeable for PanickyWriteable<T>
+where
+    T: FallibleWriteable,
+    T::Error: fmt::Debug,
+{
+    #[inline]
+    fn write_to<W: fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+        let result = self.0.fallible_write_to(sink, panicky_handler)?;
+        result.unwrap();
+        Ok(())
+    }
+
+    #[inline]
+    fn write_to_parts<S: PartsWrite + ?Sized>(&self, sink: &mut S) -> fmt::Result {
+        let result = self.0.fallible_write_to_parts(sink, panicky_handler)?;
+        result.unwrap();
+        Ok(())
+    }
+
+    #[inline]
+    fn writeable_length_hint(&self) -> LengthHint {
+        let result = self.0.fallible_writeable_length_hint(panicky_handler);
+        result.unwrap()
+    }
+
+    #[inline]
+    fn write_to_string(&self) -> Cow<str> {
+        let result = self.0.fallible_write_to_string(panicky_handler);
+        result.unwrap()
+    }
+
+    #[inline]
+    fn write_cmp_bytes(&self, other: &[u8]) -> core::cmp::Ordering {
+        let result = self.0.fallible_write_cmp_bytes(other, panicky_handler);
+        result.unwrap()
+    }
+}
+
+impl_display_with_writeable!([T] PanickyWriteable<T> where T: FallibleWriteable, T::Error: fmt::Debug);
+
+/// A writeable that panics in debug mode and writes replacements in release mode.
+///
+/// For more details, see [`TryWriteable::gigo`].
+#[derive(Debug)]
+pub struct GigoWriteable<T>(pub(crate) T);
+
+fn gigo_handler<E: fmt::Debug>(e: E) -> Result<E, Infallible> {
+    debug_assert!(false, "{e:?}");
+    Ok(e)
+}
+
+impl<T> Writeable for GigoWriteable<T>
+where
+    T: FallibleWriteable,
+    T::Error: Writeable + fmt::Debug,
+{
+    #[inline]
+    fn write_to<W: fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+        let result = self.0.fallible_write_to(sink, gigo_handler)?;
+        result.unwrap_infallible();
+        Ok(())
+    }
+
+    #[inline]
+    fn write_to_parts<S: PartsWrite + ?Sized>(&self, sink: &mut S) -> fmt::Result {
+        let result = self.0.fallible_write_to_parts(sink, gigo_handler)?;
+        result.unwrap_infallible();
+        Ok(())
+    }
+
+    #[inline]
+    fn writeable_length_hint(&self) -> LengthHint {
+        let result = self.0.fallible_writeable_length_hint(gigo_handler);
+        result.unwrap_infallible()
+    }
+
+    #[inline]
+    fn write_to_string(&self) -> Cow<str> {
+        let result = self.0.fallible_write_to_string(gigo_handler);
+        result.unwrap_infallible()
+    }
+
+    #[inline]
+    fn write_cmp_bytes(&self, other: &[u8]) -> core::cmp::Ordering {
+        let result = self.0.fallible_write_cmp_bytes(other, gigo_handler);
+        result.unwrap_infallible()
+    }
+}
+
+impl_display_with_writeable!([T] GigoWriteable<T> where T: FallibleWriteable, T::Error: Writeable, T::Error: fmt::Debug);
+
+impl<T> FallibleWriteable for Option<T>
```

Comment:
I would prefer implementing `FallibleWriteable` on `Result<T, E>`, so that the caller can control the error type. Using `Option` means the caller has to keep track of errors outside of this fallibility, because the handler will always receive the ZST.